contracts/SecurityTokenRegistry.sol

```diff
@@ -453,42 +460,107 @@ contract SecurityTokenRegistry is ISecurityTokenRegistry, EternalStorage {
+    }
+
+    /**
+     * @notice Returns the list of all tokens
+     * @dev Intention is that this is called off-chain so block gas limit is not relevant
+     */
+    function getTokens() public view returns(address[]) {
+        return _getTokens(true, address(0));
+    }
+
+    /**
+     * @notice Returns the list of tokens owned by the selected address
+     * @param _allTokens if _allTokens is true returns all tokens despite on the second parameter
+     * @param _owner is the address which owns the list of tickers
+     */
+    function _getTokens(bool _allTokens, address _owner) internal view returns(address[]) {
         // Loop over all active users, then all associated tickers of those users
         // This ensures we find tokens, even if their owner has been modified
         address[] memory activeUsers = getArrayAddress(Encoder.getKey("activeUsers"));
         bytes32[] memory tickers;
-        address token;
         uint256 count = 0;
         uint256 i = 0;
         uint256 j = 0;
         for (i = 0; i < activeUsers.length; i++) {
             tickers = getArrayBytes32(Encoder.getKey("userToTickers", activeUsers[i]));
             for (j = 0; j < tickers.length; j++) {
-                token = getAddressValue(Encoder.getKey("tickerToSecurityToken", Util.bytes32ToString(tickers[j])));
-                if (token != address(0)) {
-                    if (IOwnable(token).owner() == _owner) {
-                        count = count + 1;
-                    }
+                if (address(0) != _ownerInToken(tickers[j], _allTokens, _owner)) {
+                    count++;
                 }
             }
         }
-        uint256 index = 0;
         address[] memory result = new address[](count);
+        count = 0;
+        address token;
         for (i = 0; i < activeUsers.length; i++) {
             tickers = getArrayBytes32(Encoder.getKey("userToTickers", activeUsers[i]));
             for (j = 0; j < tickers.length; j++) {
-                token = getAddressValue(Encoder.getKey("tickerToSecurityToken", Util.bytes32ToString(tickers[j])));
-                if (token != address(0)) {
-                    if (IOwnable(token).owner() == _owner) {
-                        result[index] = token;
-                        index = index + 1;
-                    }
+                token = _ownerInToken(tickers[j], _allTokens, _owner);
+                if (address(0) != token) {
+                    result[count] = token;
+                    count++;
                 }
             }
         }
         return result;
     }
 
+    function _ownerInToken(bytes32 _ticker, bool _allTokens, address _owner) internal view returns(address) {
+        address token = getAddressValue(Encoder.getKey("tickerToSecurityToken", Util.bytes32ToString(_ticker)));
+        if (token != address(0)) {
+            if (_allTokens || IOwnable(token).owner() == _owner) {
+                return token;
+            }
+        }
+        return token;
```

Comment:
Shouldn't we return `address(0)` here?

Reply:
It will return automatically the `address(0)` I think as you read data from the eternal storage.

Reply:
I meant for the case when token exists but the owner is someone else.